proximal/examples/test_deconv_sinv_psf.py

```diff
@@ -0,0 +1,115 @@
+
+#===========
+# SOLVER
+def my_factory(psf, im_shape, fast_norm_estimate=True, normalize_psf=False):
+    # Compute the norm of the Gram matrix, i.e. max || convT(conv(x)) || for all images x.
+    scaling_factor = (np.sum(psf) if fast_norm_estimate else None)
+
+    if normalize_psf:
+        psf /= scaling_factor
+    
+    def forward(im_src, im_sim):
+        assert np.all(im_src.shape == im_sim.shape)
+        im_sim[:] = convolve(im_src, psf, mode='same', method='auto')
+
+    def adjoint(im_src, im_sim):
+        assert np.all(im_src.shape == im_sim.shape)
+        im_sim[:] = convolve(im_src, np.rot90(psf, k=2), mode='same', method='auto')
+    
+    if not fast_norm_estimate:
+        norm_bound = None
+    elif normalize_psf:
+        norm_bound = 1.0
+    else:
+        norm_bound = scaling_factor
+
+    return LinOpFactory(im_shape, im_shape, forward, adjoint, norm_bound=norm_bound)
+
+def cost_function(x, im, psf, mu=0.1, using_px_conv=True, fast_norm_estimate=True): 
+    if using_px_conv:
+        im_blur = px.conv(psf, x)
+    else:
+        blur_op = my_factory(psf, im.shape, fast_norm_estimate=fast_norm_estimate)
+        im_blur = blur_op(x)
+
+    data_term = px.sum_squares(im_blur - im) 
+    im_grad = px.grad(x)
+    grad_term = mu * px.group_norm1(im_grad, group_dims=(2, )) #+ (1 - mu) * px.sum_squares(im_grad)
+    return data_term + grad_term #+ px.nonneg(im_blur)
+
+x = px.Variable(im_sb.shape)
+
+# Expected output: convergence within 2 iterations
+# Estimated params [lambda = 1.000 | mu = 1.000 | L_est = 1.0000]
+# iter 0: ||r||_2 = 10.857, eps_pri = 14.492, ||s||_2 = 0.000, eps_dual = 10.241
+# op = LIN-ADMM iteration, evals = 1, total_time (ms) = 918.6395730357617, avg_time (ms) = 918.6395730357617
+prob = px.Problem(cost_function(x, im_sb, psf=psf_cropped, mu=0.1, using_px_conv=True))
+prob.solve(solver='ladmm', eps_abs=1e-2, verbose=True)
+plt.subplot(234); plt.imshow(x.value); plt.colorbar(); plt.title('Reconstructed Image / px.conv')
+
+# Expected output: Maximum iteration reached at 100; solver fails to converge.
+# Estimated params [lambda = 1.000 | mu = 1.000 | L_est = 1.0000]
+# iter 0: ||r||_2 = 6042.671, eps_pri = 23.779, ||s||_2 = 5546.468, eps_dual = 15.786
+# iter 20: ||r||_2 = 98.154, eps_pri = 23.784, ||s||_2 = 0.023, eps_dual = 10.329
+# iter 40: ||r||_2 = 51.718, eps_pri = 23.784, ||s||_2 = 0.007, eps_dual = 10.304
+# iter 60: ||r||_2 = 44.428, eps_pri = 23.784, ||s||_2 = 0.005, eps_dual = 10.291
+# iter 80: ||r||_2 = 44.191, eps_pri = 23.784, ||s||_2 = 0.003, eps_dual = 10.280
+# iter 100: ||r||_2 = 40.796, eps_pri = 23.784, ||s||_2 = 0.002, eps_dual = 10.277
+prob = px.Problem(cost_function(x, im_sb, psf=psf_cropped, mu=0.1, using_px_conv=False, fast_norm_estimate=False))
+prob.solve(solver='ladmm', eps_abs=1e-2, verbose=True)
+plt.subplot(235); plt.imshow(x.value); plt.colorbar(); plt.title('Reconstructed Image / custom linop')
+
+plt.show()
```

Comment:
> Also regarding the space-invariant case, I generated results using with px.conv() and then using BlackBox based convolution forward operator by toggling the using_px_conv flag, and the results are visibly bad with the latter. Do you know why?

Tagging @SteveDiamond for help. 

The question is regarding the lack of convergence when a user provided `linop` is used instead of `proximal.linop.conv`. As shown in the example, when `using_px_conv=False`, the ADMM solver struggled to converge. It happens even when @shnaqvi toggles `fast_norm_estimate=False`. PSF blur kernel normalization (line 58) is not possible due to the nature of the problem.

Another off-topic question for everyone: is there a strong reason `scipy.signal.convolve` is chosen over `scipy.ndimage.convolve`? They have different boundary conditions in the implementation.

Here's the solutions with or without custom `conv` operator:
![image](https://github.com/comp-imaging/ProxImaL/assets/12185630/768d164a-7212-40af-b509-c05202c12edf)

Reply:
Another off-topic question for everyone: is there a strong reason `scipy.signal.convolve` is chosen over `scipy.ndimage.convolve`? They have different boundary conditions in the implementation.

Reference: https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve.html#scipy.ndimage.convolve

![image](https://github.com/comp-imaging/ProxImaL/assets/12185630/01be78e3-0c54-40c6-8a78-a43702876d32)

Reply:
I've just found `scipy.ndimage.convolve` to be unbearably slow (50x slower in our case of large images, taking 160sec vs. 3sec by `scipy.signal.convolve`).

Reply:
Hmm a lot going on here, apologies if I'm missing something. I believe the default proximal convolution has circular boundary conditions, which allows for a faster and higher precision implementation of ADMM using FFTs. @shnaqvi is it possible to pad your images such that circular boundary conditions can be used?

Reply:
@antonysigma , to add to your comment to @SteveDiamond , this looks like a behavior attributed to `BlackBox` class that we're defining, inheriting from `LinOp` class. 

So even if I replace the `px.conv(psf, im)` linear operator with the following equivalent `BlackBox` implementation using the `px.conv()`, I get the same slow convergence and poorer results (fails to converge after 280 iterations) than what I get from using `px.conv` directly (converges in 2 iterations).

```Python
def my_factory(psf, im_shape, fast_norm_estimate=True):
    
    def forward(im_src, im_sim):
        im_sim[:] = px.conv(psf, im_src).value

    def adjoint(im_src, im_sim):
        im_sim[:] = px.conv(np.rot90(psf, k=2), im_src).value 

    scaling_factor = (np.sum(psf) if fast_norm_estimate else None)
    return LinOpFactory(im_shape, im_shape, forward, adjoint, norm_bound=scaling_factor)
```

<img width="986" alt="image" src="https://github.com/comp-imaging/ProxImaL/assets/4764586/305a112d-f47b-4eea-865a-e83170e020c8">

Reply:
That makes sense because proximal knows the FFT trick for ``px.conv`` but not for a blackbox operator.

Reply:
so you think its not the choice of `convolve` operator I use right or need for padding the input, but proximal's ability to not apply the FFT trick in blackbox operator? As @antonysigma stated, I want to use a custom `linop` for my space-variant forward operator doing sum of weighted convolutions. So how can I do that effectively with proximal?

Reply:
Re: direct FFT method.

@SteveDiamond Does solver `ladmm` and `pc` absorb the convolution kernel into norm-2 function by default, as shown in the screenshot below? I recalled from the example `test_deconv.py` that the PC solver compute the proximal via a direct `lsqr` or `qp` method with FFT, but yeah, I can find out by studying the `proximal.lin_op.sum_squares.least_square` code.

![image](https://github.com/comp-imaging/ProxImaL/assets/12185630/c307b270-8d4e-4702-82d4-73bbb99fa17c)


@shnaqvi Steven is referring to this paragraph for `solver='admm'`, but I recall you prefer either `pc` or `ladmm`?

![image](https://github.com/comp-imaging/ProxImaL/assets/12185630/97d30e87-2b74-4836-8538-f437cbb17622)

Reply:
That's right, I mean the absorption into the L2 norm.

Reply:
So in light of this @antonysigma, is it best to write our forward operator for the `space-variant` deconvolution case using the `px.conv` and other built-in linear operators, like `px.mul_elemwise`, `px.sum`? And in that case, is this an equivalent formulation of our forward model?

```Python
    im_blur = px.sum([px.conv(psf_modes[i], px.mul_elemwise(weights[i], x)) for i in range(n_psf)])
```

Reply:
> is this an equivalent formulation of our forward model?

Yes I guess. It doesn't hurt to prove it by trial and error. ProxImaL is designed to let you (the end user) focus on problem formulation, decoupled from the solver internal implementations.

>  is it best to write our forward operator for the space-variant deconvolution case using the px.conv and other built-in linear operators, like px.mul_elemwise, px.sum ?

Not sure if ProxImaL will be overwhelmed by 15+ `lin_op.conv` nodes in the internal representation (IR). Can you upload the code in this PR? I will also try it out at the spare time.

Or, see if you can customize your own proximal function that implements the following function with direct FFT method.

```math
\begin{align}
f(u) &= \left\Vert \sum_{j=1}^N ( F^T h_j F u ) - b \right\Vert_2^2 \\
prox(v) &= \arg\min_v f(u) + \frac{\rho}{2} \Vert u - v \Vert_2^2
\end{align}
```

> I tried this with admm and ladmm but it's not giving me sensible result.

Deferring to @SteveDiamond .

Reply:
I like @antonysigma's suggestion of implementing your own proximal operator, though I believe it would need to be for
```f(u1,...,un) = \|\sum_i F^Th_iFu_i - b_i\|^2_2```.

@shnaqvi ADMM should always converge if you let it run long enough.

Reply:
Hi @SteveDiamond , a question was raised on whether the diagonal (or block diagonal?) representation of `px.prox_fn.sum` is absorbed into `px.sum_squares` in order to utilize the direct FFT method. And, how an average user can prove it in the logs via `verbose=True`.

https://github.com/comp-imaging/ProxImaL/blob/bcc5bf457f4c48f999bc970cf0319dad8509a069/proximal/lin_ops/sum.py#L66-L70

Reply:
> though I believe it would need to be for `f(u1,...,un) = \|\sum_i F^Th_iFu_i - b_i\|^2_2`.

@SteveDiamond I like the idea, but I am not sure if the problem formulation is negotiable. Deferring to @shnaqvi for clarification. His problem is a variant of image segmentation problem, which by definition not possible to segment the source image `b` into its components ahead of time.

Reply:
> Hi @SteveDiamond , a question was raised on whether the diagonal (or block diagonal?) representation of `px.prox_fn.sum` is absorbed into `px.sum_squares` in order to utilize the direct FFT method. And, how an average user can prove it in the logs via `verbose=True`.
> 
> https://github.com/comp-imaging/ProxImaL/blob/bcc5bf457f4c48f999bc970cf0319dad8509a069/proximal/lin_ops/sum.py#L66-L70

It appears to intelligently handle a sum of convolutions. This is the key function to determine whether it can be diagonalized in the frequency domain:
https://github.com/comp-imaging/ProxImaL/blob/644ad376d8d2deedb6ab2daa73e95ba5978bbbd9/proximal/lin_ops/sum.py#L54

Reply:
Thank you @SteveDiamond for the quick note. It appears for absorption of `conv` into `sum_squares` to happen, the problem formulation must be `px.sum( px.conv(h[i], x) + ...)`, not `px.sum( px.conv(h[i], elemwise_mul(w[i], x)+ ...)`.

Is my understanding correct? The function in `proximal.algorithms.invert.get_diag_quads()` seems to require that. That is, absorption does not happen for Shnaqvi's problem formulation.

Reply:
That is correct. I also don't believe absorption is even possible in Shnaqvi's problem formulation. Elementwise weights are not a diagonal matrix in the frequency domain I believe.

Reply:
I see. @shnaqvi and I were confused at the following description in the publication. Thanks again for the clarifications!

> Our compiler by default iterates over the proxable functions, and for each one repeatedly absorbs linear operators until doing so would substantially increase the cost of the proximal operator.

Reply:
> I see. @shnaqvi and I were confused at the following description in the publication. Thanks again for the clarifications!
> 
> > Our compiler by default iterates over the proxable functions, and for each one repeatedly absorbs linear operators until doing so would substantially increase the cost of the proximal operator.

I think the meaning there was that in theory you could absorb everything into the proximal operator and evaluate it using a generic optimization method. But that would go against the motivation of proximal algorithms, which is to use only proximal operators that have an efficient way to evaluate them.

Reply:
BTW, I love how you have continued this project and improved it so much! Let me know if you want to chat about any big picture aspects. The cvxpy discord is good place to chat https://discord.gg/4urRQeGBCr